src/pg_query_summary.c

```diff
@@ -0,0 +1,761 @@
+				//  optimization opportunity available.
+				state->save_filter_columns = true;
+				pg_query_summary_walk_impl((Node*) stmt->whereClause, state);
+				state->save_filter_columns = false;
+			}
+
+			break;
+		}
+
+		case T_InsertStmt: {
+			add_from_clause_item(castNode(InsertStmt, node)->relation, CONTEXT_DML, state);
+			break;
+		}
+
+		case T_UpdateStmt: {
+			UpdateStmt *stmt = castNode(UpdateStmt, node);
+			add_from_clause_item(stmt->relation, CONTEXT_DML, state);
+			add_from_clause_items(stmt->fromClause, CONTEXT_SELECT, state);
+			break;
+		}
+
+		case T_DeleteStmt: {
+			DeleteStmt *stmt = castNode(DeleteStmt, node);
+			add_from_clause_item(stmt->relation, CONTEXT_DML, state);
+			add_from_clause_items(stmt->usingClause, CONTEXT_SELECT, state);
+			break;
+		}
+
+		case T_CopyStmt: {
+			CopyStmt* stmt = castNode(CopyStmt, node);
+			add_from_clause_item(stmt->relation, CONTEXT_SELECT, state);
+			return false; // Not handled by tree walker.
+		}
+
+		case T_AlterTableStmt: {
+			AlterTableStmt* stmt = castNode(AlterTableStmt, node);
+
+			// `ALTER INDEX index_name` is ignored.
+			if (stmt->objtype != OBJECT_INDEX)
+				add_from_clause_item(stmt->relation, CONTEXT_DDL, state);
+
+			return false; // Not handled by tree walker.
+		}
+
+		case T_CreateStmt:
+			add_from_clause_item(castNode(CreateStmt, node)->relation, CONTEXT_DDL, state);
+			return false; // Not handled by tree walker.
+
+		case T_CreateTableAsStmt: {
+			CreateTableAsStmt* stmt = castNode(CreateTableAsStmt, node);
+			if (stmt->into && stmt->into->rel)
+				add_from_clause_item(stmt->into->rel, CONTEXT_DDL, state);
+
+			if (stmt->query)
+				pg_query_summary_walk_impl(stmt->query, state);
+
+			return false; // Not handled by tree walker.
+		}
+
+		case T_TruncateStmt:
+			add_from_clause_items(castNode(TruncateStmt, node)->relations, CONTEXT_DDL, state);
+			return false; // Not handled by tree walker.
+
+		case T_ViewStmt: {
+			ViewStmt* stmt = castNode(ViewStmt, node);
+			add_from_clause_item(stmt->view, CONTEXT_DDL, state);
+			pg_query_summary_walk_impl(stmt->query, state);
+			return false; // Not handled by tree walker.
+		}
+
+		case T_IndexStmt: {
+			IndexStmt* stmt = castNode(IndexStmt, node);
+
+			add_from_clause_item(stmt->relation, CONTEXT_DDL, state);
+
+			ListCell* lc = NULL;
+			foreach(lc, stmt->indexParams) {
+				IndexElem* index_elem = lfirst(lc);
+				if (index_elem->expr)
+					pg_query_summary_walk_impl(index_elem->expr, state);
+			}
+
+			if (stmt->whereClause)
+				pg_query_summary_walk_impl(stmt->whereClause, state);
+
+			return false; // Not handled by tree walker.
+		}
+
+		case T_CreateTrigStmt:
+			add_from_clause_item(castNode(CreateTrigStmt, node)->relation, CONTEXT_DDL, state);
+			return false; // Not handled by tree walker.
+
+		case T_RuleStmt: {
+			add_from_clause_item(castNode(RuleStmt, node)->relation, CONTEXT_DDL, state);
+			return false; // Not handled by tree walker.
+		}
```

Comment:
```suggestion
		case T_RuleStmt:
			add_from_clause_item(castNode(RuleStmt, node)->relation, CONTEXT_DDL, state);
			return false; // Not handled by tree walker.
```

(since it doesn't define a variable)